File: school-admin-web/src/lib/utils/password-policy.ts
// Password policy for admin-set staff passwords (A2). Mirrors the Firebase Auth
// console policy — 14+ chars with an upper, lower, digit and special character.
// Passwords set via the Admin SDK (createUser) bypass the console policy, so
// this is the server-side gate for the portal's "Add Staff" flow; it also drives
// the client modal's inline validation so the two never drift.
export const PASSWORD_MIN_LENGTH = 14;

/** The first unmet requirement as a human phrase, or null if the password is strong. */
export function passwordIssue(pw: string): string | null {
  if (pw.length < PASSWORD_MIN_LENGTH) return `at least ${PASSWORD_MIN_LENGTH} characters`;
  if (!/[A-Z]/.test(pw)) return 'an uppercase letter';
  if (!/[a-z]/.test(pw)) return 'a lowercase letter';
  if (!/[0-9]/.test(pw)) return 'a number';
  if (!/[^A-Za-z0-9]/.test(pw)) return 'a special character';
  return null;
}

export function isStrongPassword(pw: string): boolean {
  return passwordIssue(pw) === null;
}

export const PASSWORD_REQUIREMENT_TEXT =
  `Must be at least ${PASSWORD_MIN_LENGTH} characters and include an uppercase letter, a lowercase letter, a number and a special character.`;

File: lib/core/validation/password_policy.dart
/// Shared password policy for non-student accounts (A2).
///
/// Mirrors the Firebase Auth console policy and the web portal validator: at
/// least 14 characters with an uppercase letter, a lowercase letter, a number
/// and a special character. Keeping one source of truth stops the rule drifting
/// between the parent-signup, teacher-signup and school-onboarding screens
/// (which each had their own inline 8-char check).
library;

const int kMinPasswordLength = 14;

/// The first unmet requirement as a short phrase (e.g. 'a special character'),
/// or null if [value] satisfies the policy.
String? passwordIssue(String value) {
  if (value.length < kMinPasswordLength) {
    return 'at least $kMinPasswordLength characters';
  }
  if (!RegExp(r'[A-Z]').hasMatch(value)) return 'an uppercase letter';
  if (!RegExp(r'[a-z]').hasMatch(value)) return 'a lowercase letter';
  if (!RegExp(r'[0-9]').hasMatch(value)) return 'a number';
  if (!RegExp(r'[^A-Za-z0-9]').hasMatch(value)) return 'a special character';
  return null;
}

/// Whether [value] satisfies the full password policy.
bool isPasswordCompliant(String value) => passwordIssue(value) == null;

/// One-line requirement text for helper/hint copy.
const String kPasswordRequirementText =
    'At least $kMinPasswordLength characters with an uppercase and lowercase '
    'letter, a number and a special character.';
